dll/win32/msafd/misc/sndrcv.c

```diff
@@ -248,35 +304,40 @@ WSPRecv(SOCKET Handle,
         if ((Socket->SharedData->CreateFlags & SO_SYNCHRONOUS_NONALERT) != 0)
         {
             TRACE("Opened without flag WSA_FLAG_OVERLAPPED. Do nothing.\n");
-            return MsafdReturnWithErrno(0, lpErrno, 0, lpNumberOfBytesRead);
+            NtClose(SockEvent);
+            HeapFree(GlobalHeap, 0, RecvInfo);
+            return MsafdReturnWithErrno(STATUS_SUCCESS, lpErrno, 0, lpNumberOfBytesRead);
         }
-        if (lpCompletionRoutine == NULL)
+
+        APCContext = HeapAlloc(GlobalHeap, 0, sizeof(AFDRECVAPCCONTEXT));
+        if (!APCContext)
         {
-            /* Using Overlapped Structure, but no Completion Routine, so no need for APC */
-            APCContext = lpOverlapped;
-            APCFunction = NULL;
-            Event = lpOverlapped->hEvent;
+            ERR("Not enough memory for APC Context\n");
+            NtClose(SockEvent);
+            HeapFree(GlobalHeap, 0, RecvInfo);
+            return MsafdReturnWithErrno(STATUS_INSUFFICIENT_RESOURCES, lpErrno, 0, lpNumberOfBytesRead);
```

Comment:
What should happen to the `*ReceiveFlags` in these error code paths in this WSPRecv function?
Should they be left unmodified?

Reply:
Not sure, the code is the same as original.